File: solutions/Javascript/Word Break.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */

var wordBreak = function (s, wordDict) {
  return Break(s, 0, wordDict, {});
};

var Break = function (s, pos, words, memo) {
  if (pos === s.length) return true;
  if (memo[pos]) return memo[pos];
  for (let i = pos; i < s.length; i++) {
    if (
      words.indexOf(s.substr(pos, i - pos + 1)) > -1 &&
      Break(s, i + 1, words, memo)
    ) {
      return (memo[pos] = true);
    }
  }
  return (memo[pos] = false);
};

File: solutions/cpp/Word Break.cpp
#include <vector>
#include <string>
#include <map>
#include <set>
using namespace std;
class Solution
{
    map<int, bool> memo;
    set<string> words;

public:
    bool wordBreak(string s, vector<string> &wordDict)
    {
        for (auto word : wordDict)
            words.insert(word);
        return Break(s, 0);
    }
    bool Break(string s, int pos)
    {
        if (pos == s.length())
            return true;
        if (memo.count(pos))
            return memo[pos];
        for (int i = pos; i < s.length(); i++)
        {
            if (words.count(s.substr(pos, i - pos + 1)) > 0 && Break(s, i + 1))
                return memo[pos] = true;
        }
        return memo[pos] = false;
    }
};
